src/librustc_target/abi/call/riscv.rs

```diff
@@ -1,49 +1,341 @@
+        return None;
+    }
+    match (field1_kind, field2_kind) {
+        (FieldKind::Float, FieldKind::Integer) | (FieldKind::Integer, FieldKind::Float) => {
+            Some(FloatArgPass::MixedPair)
+        }
+        (FieldKind::Float, FieldKind::Float) => Some(FloatArgPass::FloatPair),
+        (FieldKind::Float, FieldKind::Unknown) => Some(FloatArgPass::Float),
+        _ => None,
+    }
+}
+
+fn classify_ret<'a, Ty, C>(cx: &C, arg: &mut ArgAbi<'a, Ty>, xlen: u64, flen: u64)
+where
+    Ty: TyLayoutMethods<'a, C> + Copy,
+    C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>>,
+{
+    if should_use_fp_conv(cx, &arg.layout, xlen, flen).is_some() {
+        return;
+    }
+
+    let total = arg.layout.size;
 
-fn classify_ret<Ty>(arg: &mut ArgAbi<'_, Ty>, xlen: u64) {
     // "Scalars wider than 2✕XLEN are passed by reference and are replaced in
     // the argument list with the address."
     // "Aggregates larger than 2✕XLEN bits are passed by reference and are
     // replaced in the argument list with the address, as are C++ aggregates
     // with nontrivial copy constructors, destructors, or vtables."
-    if arg.layout.size.bits() > 2 * xlen {
-        arg.make_indirect();
+    if total.bits() > 2 * xlen {
+        // We rely on the LLVM backend lowering code to lower passing a scalar larger than 2*XLEN.
+        if is_aggregate(arg) {
+            arg.make_indirect();
+        }
+        return;
+    }
+
+    let xlen_reg = match xlen {
+        32 => Reg::i32(),
+        64 => Reg::i64(),
+        _ => unreachable!("Unsupported XLEN: {}", xlen),
+    };
+    if is_aggregate(arg) {
+        arg.cast_to(Uniform { unit: xlen_reg, total });
+        return;
     }
 
     // "When passed in registers, scalars narrower than XLEN bits are widened
     // according to the sign of their type up to 32 bits, then sign-extended to
     // XLEN bits."
-    arg.extend_integer_width_to(xlen); // this method only affects integer scalars
+    extend_integer_width(arg, xlen);
 }
 
-fn classify_arg<Ty>(arg: &mut ArgAbi<'_, Ty>, xlen: u64) {
+fn classify_arg<'a, Ty, C>(
+    cx: &C,
+    arg: &mut ArgAbi<'a, Ty>,
+    xlen: u64,
+    flen: u64,
+    is_vararg: bool,
+    avail_gprs: &mut u64,
+    avail_fprs: &mut u64,
+) where
+    Ty: TyLayoutMethods<'a, C> + Copy,
+    C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>>,
+{
+    if !is_vararg {
+        match should_use_fp_conv(cx, &arg.layout, xlen, flen) {
+            Some(FloatArgPass::Float) if *avail_fprs >= 1 => {
+                *avail_fprs -= 1;
+                return;
+            }
+            Some(FloatArgPass::FloatPair) if *avail_fprs >= 2 => {
+                *avail_fprs -= 2;
+                return;
+            }
+            Some(FloatArgPass::MixedPair) if *avail_fprs >= 1 && *avail_gprs >= 1 => {
+                *avail_gprs -= 1;
+                *avail_fprs -= 1;
+                return;
+            }
+            _ => (),
+        }
```

Comment:
Is this relying on LLVM entirely for these cases? I would expect you would use casts (like `x86_64`) to make sure that whatever aggregate is turned into the right types for LLVM to work with, i.e.:
https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_target/abi/call/x86_64.rs#L148-L160